Support benchmarking with GHCJS

harendra-kumar opened the pull request:
I have disabled analysis support for GHCJS as it requires some C functions for
statistics. Though we can use javascript versions of those I did not spend the
effort to do that.

There is no CI script but it can be built using:

```
cabal new-bench --project-file cabal.project.ghcjs
```

We can then run the benchmarking executable manually using "node" or using a
browser:

```
node /vol/hosts/cueball/workspace/github/hs-gauge/dist-newstyle/build/x86_64-linux/ghcjs-8.6.0.1/gauge-0.2.4/b/self/build/self/self.jsexe/all.js
```

harendra-kumar committed:
Support benchmarking with GHCJS

I have disabled analysis support for GHCJS as it requires some C functions for
statistics. Though we can use javascript versions of those I did not spend the
effort to do that.

There is no CI script but it can be built using:

cabal new-bench --project-file cabal.project.ghcjs

We can then run the benchmarking executable manually using "node" or using a
browser:

node /vol/hosts/cueball/workspace/github/hs-gauge/dist-newstyle/build/x86_64-linux/ghcjs-8.6.0.1/gauge-0.2.4/b/self/build/self/self.jsexe/all.js

harendra-kumar committed:
update changelog (ghcjs)

harendra-kumar commented:
@vincenthz do you have any comments or feedback on this?

vincenthz commented on Gauge/Source/GC.hs:
```diff
@@ -43,10 +43,14 @@ supported = unsafePerformIO (readIORef supportedVar)
 supportedVar = unsafePerformIO $ do
+#if __GHCJS__
+    let b = False
+#else
```
can probably save one wrapping by switching this #else to #elif

harendra-kumar replied:
I will fix that.

vincenthz commented on Gauge/Source/Time.hsc:
```diff
@@ -71,6 +76,58 @@ withMetrics f = allocaBytes (sizeTimeRecord * 2) $ \ptr -> do
+
+instance Storable CTimespec where
+  sizeOf _ = 8
+  alignment _ = 4
```
both sizeOf and alignment strikes me as probably working in GHCJS but wrong otherwise. alignment should be 8 and sizeOf = sizeOf CTime + sizeOf CLong

harendra-kumar replied:
I copied this from GHCJS. I am not sure why they are hardcoded. I think `sizeOf = sizeOf CTime + sizeOf CLong` should be fine. I guess the right alignment would be `max (alignment  CTime) (alignment CLong)`. CTime and CLong are perhaps both 4 bytes on GHCJS and therefore alignment of 4 works. `CTimespec` is being read as a 64-bit value in the javascript code handling this.

harendra-kumar replied:
Even `sizeOf CTime + sizeOf CLong` may not be correct in general as there may be padding added due to alignment. So we should also take into account the alignment of each field.

vincenthz commented on Gauge/Source/Time.hsc:
```diff
@@ -71,6 +76,58 @@ withMetrics f = allocaBytes (sizeTimeRecord * 2) $ \ptr -> do
+  alignment _ = 4
+  peek p = do
+    s  <- peekByteOff p 0
+    ns <- peekByteOff p 4
```
same here, `4` should be replaced by `sizeOf CTime`

harendra-kumar replied:
`sizeOf CTime` may not be correct offset in general because of alignment.

vincenthz commented on Gauge/Source/Time.hsc:
```diff
@@ -71,6 +76,58 @@ withMetrics f = allocaBytes (sizeTimeRecord * 2) $ \ptr -> do
+    return (MkCTimespec s ns)
+  poke p (MkCTimespec s ns) = do
+    pokeByteOff p 0 s
+    pokeByteOff p 4 ns
```
and here

vincenthz commented on Gauge/Source/Time.hsc:
```diff
@@ -71,6 +76,58 @@ withMetrics f = allocaBytes (sizeTimeRecord * 2) $ \ptr -> do
+initialize = return ()
+
+getCycles :: IO (Cycles 'Absolute)
+getCycles = undefined
```
return some kind of 0 instead might be safer ?

harendra-kumar replied:
getting cycles is not supported on javascript, so we better know if this is ever called. Maybe we can define it as `error " getCycles not supported on GHCJS"`. Anyway, currently this is getting called only when the GAUGE_MEASURE_TIME_NEW macro is not defined. We should probably remove that macro altogether.

vincenthz commented on Gauge/Source/Time.hsc:
```diff
@@ -71,6 +76,58 @@ withMetrics f = allocaBytes (sizeTimeRecord * 2) $ \ptr -> do
+    return $ fromIntegral sec + (fromIntegral nsec / 1000000000)
+
+getCPUTime :: IO Double
+getCPUTime = undefined
```
same here, some 0.0 ?

vincenthz commented on cabal.project.ghcjs:
```diff
@@ -0,0 +1,7 @@
+source-repository-package
+  type: git
+  location: https://github.com/ghcjs/ghcjs-base
+  tag: b8d51f65ae1921b2f031710bf75e17f216de442a
```
is this file important ? why is a specific tag hardcoded here ?

harendra-kumar replied:
This just got left behind, I will remove this.

vincenthz reviewed (commented):
Not a fan of ghcjs (I don't think the approach is sustainable), but if this is useful, it doesn't seems too much hassle to add support at the moment.

harendra-kumar committed:
Address review comments

harendra-kumar commented:
@vincenthz I have addressed the review comments except the `Storable` instance. I do not understand the Haskell to JavaScript FFI/Marshalling well enough so I left it as it is, I am not sure whether whatever changes we make would make any real difference.

harendra-kumar commented:
@vincenthz is this ok to merge it now?

harendra-kumar commented:
@vincenthz can we merge this?
